### PR TITLE
Add basic means of concatenating datasets

### PR DESCRIPTION
**Pull request by alexander-soare**
## What this does

Adds a way of concatenating datasets via the training script and configuration.

You can now provide a list of datasets for the `dataset_repo_id`. Here is the documentation from `default.yaml` verbatim:

```
# You may provide a list of datasets here. `train.py` creates them all and concatenates them. Note: only data
# keys common between the datasets are kept. Each dataset gets and additional transform that inserts the
# "dataset_index" into the returned item. The index mapping is made according to the order in which the
# datsets are provided.
```

Side changes:
- Since I added `consolidate_stats` I decided to put it and `compute_stats` into one file under `lerobot/common/data/compute_stats.py`. This is to make sure to have the logic that deals with the stats structure in one place.
- The single dataset setting also uses a ConcatDataset to make the API to the `offline_dataset` object consistent downstream.
- Added the ability to disable/enable keys in LeRobotDataset.

## How it was tested

Added CI testing for `consolidate_stats` and for enable/disabling data keys.

## How to checkout & try? (for the reviewer)

Grab [act_aloha_all.txt](https://github.com/huggingface/lerobot/files/15433359/act_aloha_all.txt) and put it into `lerobot/configs/policy` as `act_aloha_all.yaml` (change the file extension from txt to yaml).

Run

```bash
python lerobot/scripts/train.py env=aloha env.task=AlohaTransferCube-v0 policy=act_aloha_all
```

- - -
This change is [Reviewable](https://reviewable.io/reviews/huggingface/lerobot/208)

### CODE REVIEW

File: lerobot/common/datasets/compute_stats.py [hunk 1]
```diff
@@ -159,3 +156,50 @@ def create_seeded_dataloader(dataset, batch_size, seed):
...
+    for dataset in ls_datasets:
+        data_keys.update(dataset.stats.keys())
+    stats = {k: {} for k in data_keys}
+    for data_key in data_keys:
+        for stat_key in ["min", "max"]:
+            stats[data_key][stat_key] = einops.reduce(
+                torch.stack([d.stats[data_key][stat_key] for d in ls_datasets if data_key in d.stats], dim=0),
+                "n ... -> ...",
+                stat_key,
+            )
```
Comment (Cadene) [inline lerobot/common/datasets/compute_stats.py:179]:
```suggestion
        for stat_key in ["min", "max"]:
            # compute `max(dataset_0["max"], dataset_1["max"], ...)`
            ds_stats = torch.stack([d.stats[data_key][stat_key] for d in ls_datasets if data_key in d.stats], dim=0)
            stats[data_key][stat_key] = einops.reduce(ds_stats, "n ... -> ...", stat_key)
```

File: lerobot/scripts/train.py [hunk 2]
```diff
@@ -248,13 +253,58 @@ def train(cfg: DictConfig, out_dir: str | None = None, job_name: str | None = No
...
+        id_to_index = {repo_id: i for i, repo_id in enumerate(cfg.dataset_repo_id)}
+        logging.info(
+            "Multiple datasets were provided. Applying the following index mapping to the provided datasets: "
+            f"{pformat(id_to_index, indent=2)}"
+        )
+        for dataset_repo_id, dataset in zip(cfg.dataset_repo_id, _offline_datasets, strict=True):
+
+            def transform(item, d_id=dataset_repo_id):
+                item["dataset_index"] = torch.tensor(id_to_index[d_id])
+                return item
```
Comment (Cadene) [inline lerobot/scripts/train.py:288]: We should probably avoid using transform for this use case. That's why I would prefer to use a `MultiLeRobotDataset` class that would contain the mapping in its attribute and add the `dataset_index` key in its getitem.
What do you think?

File: lerobot/common/datasets/compute_stats.py [hunk 3]
```diff
@@ -159,3 +156,50 @@ def create_seeded_dataloader(dataset, batch_size, seed):
             "min": min[key],
         }
     return stats
+
+
+def consolidate_stats(ls_datasets: list[LeRobotDataset]) -> dict[str, torch.Tensor]:
+    """Consolidate stats across multiple datsets into one set of stats without recomputing from scratch.
+
+    This is technically not possible to do for the standard deviation as the sum of squares is required.
+    Instead, we choose the
```
[the remark this answers is not in the document]
Comment (alexander-soare) [reply lerobot/common/datasets/compute_stats.py:165]: I actually removed this paragraph as it **is** technically possible.

File: lerobot/common/datasets/lerobot_dataset.py [hunk 4]
```diff
@@ -65,6 +66,23 @@ def __init__(
...
+                raise ValueError(f"Provided key {k} is not present in the dataset.")
+            self.disabled_data_keys.add(k)
+
+    def enable_data_keys(self, keys: list[str]):
+        """Enable selected data keys in the __getitem__ return."""
+        for k in keys:
+            if k not in self.hf_dataset.features:
+                raise ValueError(f"Provided key {k} is not present in the dataset.")
+            if k in self.disabled_data_keys:
+                self.disabled_data_keys.remove(k)
```
[the remark this answers is not in the document]
Comment (alexander-soare) [reply lerobot/common/datasets/lerobot_dataset.py:85]: I'm actually not sure what you mean, but we can take it offline.

File: lerobot/common/datasets/lerobot_dataset.py [hunk 5]
```diff
@@ -183,7 +205,12 @@ def from_preloaded(
         stats=None,
         info=None,
         videos_dir=None,
-    ):
+    ) -> "LeRobotDataset":
+        """Create a LeRobot Dataset from existing data and attributes instead of loading from the filesystem.
+
```
[the remark this answers is not in the document]
Comment (alexander-soare) [reply lerobot/common/datasets/lerobot_dataset.py:210]: Done.

File: lerobot/scripts/train.py [hunk 6]
```diff
@@ -337,34 +390,13 @@ def evaluate_and_checkpoint_if_needed(step):
...
-    )
-    dataloader = torch.utils.data.DataLoader(
-        concat_dataset,
-        num_workers=4,
-        batch_size=cfg.training.batch_size,
-        sampler=sampler,
-        pin_memory=device.type != "cpu",
-        drop_last=False,
-    )
-
```
[the remark this answers is not in the document]
Comment (alexander-soare) [reply lerobot/scripts/train.py:408]: I realized I forgot to remove it in the PR where I removed online training. All this code is hanging around here doing nothing.

File: lerobot/common/datasets/compute_stats.py (hunk 1 above)
Comment (alexander-soare) [reply lerobot/common/datasets/compute_stats.py:179]: Done. Good suggestion.

File: lerobot/scripts/train.py (hunk 2 above)
Comment (alexander-soare) [reply lerobot/scripts/train.py:318]: Agreed. I will wrap up your comments and start a separate PR with the `MultiLeRobotDatset` approach.

File: lerobot/common/datasets/lerobot_dataset.py [hunk 7]
```diff
@@ -65,6 +66,23 @@ def __init__(
...
+        self.disabled_data_keys: set[str] = set()
+
+    def disable_data_keys(self, keys: Iterable[str]):
+        """Disable selected data keys in the __getitem__ return."""
+        for k in keys:
+            if k not in self.hf_dataset.features:
+                raise ValueError(f"Provided key {k} is not present in the dataset.")
+            self.disabled_data_keys.add(k)
+
+    def enable_data_keys(self, keys: list[str]):
```
[the remark this answers is not in the document]
Comment (alexander-soare) [reply lerobot/common/datasets/lerobot_dataset.py:79]: Agreed. I will wrap up your comments and start a separate PR with the MultiLeRobotDatset approach.

**Comment by alexander-soare** [comment]
Covered by MultiLerobotDataset

**Pull request closed without merging**